ostd/src/arch/x86/allocator.rs

```diff
@@ -0,0 +1,63 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::{borrow::ToOwned, vec::Vec};
+
+use align_ext::AlignExt;
+
+use crate::{boot::memory_region::MemoryRegionType, io::io_mem::allocator::IoMemAllocatorBuilder};
+
+/// Initializes the allocatable MMIO area based on the x86-64 memory distribution map.
+///
+/// In x86-64, the available physical memory area is divided into two regions below 32 bits (Low memory)
+/// and above (High memory). The area from the top of low memory to 0xffff_ffff and the area after the
+/// top of high memory are available MMIO areas.
+pub(super) fn construct_io_mem_allocator_builder() -> IoMemAllocatorBuilder {
+    // TODO: Add MMIO regions below 1MB (e.g., VGA framebuffer).
+    let reversed_regions = {
+        let mut regions = crate::boot::memory_regions().to_owned();
+        assert!(!regions.is_empty());
+        regions.sort_by_key(|a| a.base());
+        regions.reverse();
+        regions
+    };
+
+    let mut ranges = Vec::with_capacity(2);
+    let mut is_tohm_found = false;
+
+    const LOW_MMIO_TOP: usize = 0x1_0000_0000;
+    const LOW_MMIO_ALIGN: usize = 0x1000_0000;
+    const HIGH_MMIO_TOP: usize = 0x8000_0000_0000;
+    const HIGH_MMIO_ALIGN: usize = 0x1_0000_0000;
+
+    for region in reversed_regions {
+        if region.typ() != MemoryRegionType::Usable {
+            continue;
```

Comment:
Yes, we shouldn't ignore the other regions like `BadMemory` and `NonVolatileSleep`. However, I think we should ignore the `Reserved` region since the MMIO regions for PCIe and APIC devices (VirtIO's BARs may also included) are marked as `Reserved`.

The `Reserved` region contains different types of areas. But I think the current implementation can only reach the MMIO region in `Reserved` since it searches only two "holes": below 32 bits (Low memory) and above (High memory). Also, the alignment (0x1000_0000 and 0x1_0000_0000) will "ignore" the `Reserved` regions that access physical memory.

For example, this region will only include the PCIe MMIO region, disregarding the `Reserved` section between `NonVolatileSleep` and `Usable` (This is an example from an instance, and I have checked the supported protocols in Asterinas-x86_64, which also discard the non-MMIO `Reserved` field.) :

```text
    MemoryRegion {
        base: 0xb0000000,
        len: 0x10000000,
        typ: Reserved,
    },
    MemoryRegion {
        base: 0x7ff78000,
        len: 0x88000,
        typ: NonVolatileSleep,
    },
    MemoryRegion {
        base: 0x7ff58000,
        len: 0x20000,
        typ: Reserved,
    },
    MemoryRegion {
        base: 0x7fbff000,
        len: 0x359000,
        typ: Usable,
    },
```

Reply:
What about the 3.5G to 4GB memory region?

On many 32-bit systems (and even on some 64-bit boards for compatibility), the top portion of the 4 GB address space—often somewhere in the [3 GB..4 GB) range—gets reserved for PCI/PCIe MMIO and chipset registers. A typical layout might look like:

* 0xFEC0_0000 – 0xFECF_FFFF: I/O APIC area (exact size can vary, but 1 MB is common).
* 0xFEE0_0000 – 0xFEEF_FFFF: Local APIC area.
* 0xFED0_0000 – 0xFEDx_xxxx: HPET or other chipset registers (addresses vary depending on chipset).
* 0xFEEx_xxxx – 0xFFFF_FFFF: Additional chipset or BIOS areas. On some systems, the top of physical memory (0xFFF0_0000 – 0xFFFF_FFFF) may be reserved for a BIOS flash chip or System Management Mode (SMM) firmware.